Added stacktrace logging for failed simulations

kalouantonis opened the pull request:
Currently there's no support for this feature and it was becoming quite unwieldy for me to use in larger simulations.

I could also change this PR to output to `results/<sim-name>/input/errors.log`

mhjort commented:
Cool. Thanks for the PR. This is definitely an useful feature. As you suggested it would be better to write this to a file. Instead of input directory the path could be ```results/<sim-name>/errors.log```.

While trying out this PR I found out that there was a bug in test code related to exception testing. That is now fixed in ```master``` branch. You can rebase your PR against to that. Then you should be able to run tests successfully.

For QA purposes I have been trying to keep test coverage for ```simulation``` namespace high. Could you add an assertion to ```when-function-returns-exception-it-is-handled-as-ko``` test case for stack trace printing? Something simple like first deleting the file before the test case and then asserting the file is not empty after the test run would be enough.

kalouantonis commented:
I've updated the code quite a bit to get `results-dir` passed to the relevant simulation functions and I've written a new test (and a fixture) `when-function-throws-exception-it-is-logged` that just checks that the file exists and contains the correct exception message.

I've added a `FIXME` [here](https://github.com/mhjort/clj-gatling/pull/45/files#diff-2880d301effb60bf701bf31b6d3754b0R219) because I'm (almost) sure that output will be jumbled up when using concurrency > 1 but I haven't actually tested this. IMO, a way of dealing with this is just to use a proper logging library like [timbre](https://github.com/ptaoussanis/timbre) rather than trying to do it ourselves.

Lastly, I updated the `README` section on error logging.

Let me know if I've missed anything.

kalouantonis commented:
Also, I don't know why the CI is failing and I can't seem to replicate it on my system. Is there a way to run it again?

mhjort commented:
CI failure was a weird one. I ran it again and it passes now. If this happens again I will investigate that particular test case in more detail.

mhjort commented on test/clj_gatling/simulation_test.clj:
```diff
@@ -195,6 +211,16 @@
+  (let [result (-> {:name "Exception logging scenario"
+                    :steps [{:name "Throwing" :request (fn [_] (throw (Exception. "Simulated")))}]}
+                   ;; FIXME: output will probably be jumbled with concurrency > 1
+                   (run-single-scenario :concurrency 1))]
```
I tested this with concurrency 200 and it works just fine.

mhjort commented on test/clj_gatling/simulation_test.clj:
```diff
@@ -15,11 +16,24 @@
+  (f))
+
+;; Just setup the file path once
+(use-fixtures :once setup-error-file-path)
```
Maybe no need for that comment since code is self-explanatory in this case.

mhjort commented on test/clj_gatling/simulation_test.clj:
```diff
@@ -15,11 +16,24 @@
+
+(defn- setup-error-file-path
+  [f]
+  (let [file (clojure.java.io/file error-file-path)]
```
Use require

mhjort commented on test/clj_gatling/simulation_test.clj:
```diff
@@ -15,11 +16,24 @@
+(def error-file-path "target/test-results/error.log")
+
+(defn- setup-error-file-path
+  [f]
```
Style issue. I have put function parameters to same line with function name.

mhjort commented on test/clj_gatling/simulation_test.clj:
```diff
@@ -195,6 +211,16 @@
 
+(deftest when-function-throws-exception-it-is-logged
+  ;; delete previous log data
+  (clojure.java.io/delete-file error-file-path)
```
Use require

mhjort commented on src/clj_gatling/simulation_util.clj:
```diff
@@ -3,16 +3,40 @@
+  "Append `contents` to file at `path`. The file is created
+  if it doesn't already exist."
+  [^String path contents]
+  (with-open [w (clojure.java.io/writer path :append true)]
```
Use require

mhjort commented on src/clj_gatling/core.clj:
```diff
@@ -31,29 +31,35 @@
                                           :timeout-in-ms step-timeout
-                                          :context (:context options)}
+                                          :context (:context options)
+                                          :error-file (or (:error-file options)
```
Could you document this to README?

mhjort commented:
I added few minor comments about code. Otherwise this looks really good. After those minor changes I can merge this PR. 

As you mentioned later on `timbre` could be a good option for logging. However, I wouldn't do it yet. At the moment your simple solution is robust enough.

kalouantonis commented:
Ok, I've updated everything. 

Since you've tested that logging doesn't break for a large concurrency value I also agree that `timbre` is overkill.

mhjort commented:
Looks good. Could you cleanup a git history a bit (like use fixup for `Accidentaly put a require...` commit) and then I can merge this PR.

kalouantonis commented:
I hope this is what you meant. I'm unfamiliar with rebasing and squashing

mhjort commented:
Not exactly. But no worries. I can do the git history cleanup now.

mhjort commented:
I did rebasing and squashing manually and merged that branch to master. Also released version `0.8.6` with this support. Thank you for this feature.
